### ISSUE TITLE
[BUG] Redeclaring variable in pure python mode is not raising error

### ISSUE THREAD

**Issue by matusvalo** [cython/cython#5059]
### Describe the bug

Redeclaring variable in pure python mode is behaving not consistently with cython code. E.g. in cython following code raises error:
```cython
def foo():
    cdef float a = 3.0
    cdef int a = 1
    print(a)
```

```
Error compiling Cython file:
------------------------------------------------------------
...
# cdef int a = 1
# cdef int a = 2.0

def foo():
    cdef float a = 3.0
    cdef int a = 1
             ^
------------------------------------------------------------

bug.pyx:8:13: 'a' redeclared
```

### Code to reproduce the behaviour:

```cython
def foo():
    a: cython.float = 3.0
    a: cython.int = 1
    print(a)
```

```python
$ python
Python 3.9.12 (main, May  8 2022, 18:05:13)
[Clang 12.0.0 (clang-1200.0.32.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import bug
>>> bug.foo()
1.0
>>>
```

### Expected behaviour

There are two possible solutions:
1. compiling code will fail with error as in cython code
2. code prints value `1`

### Environment

OS: macOS
Python version 3.9.12
Cython version master

**Comment by da-woods** [comment]
I think that's expected. It's legal (if poor) Python code so I think it'll just ignore contradictory annotations. I'd expect a warning though

**Comment by matusvalo** [comment]
> I think that's expected. It's legal (if poor) Python code so I think it'll just ignore contradictory annotations. I'd expect a warning though

Fair enough but in this case the code should print `1` not `1.0`. E.g. in python you got:

```python
>>> a: float = 1.0
>>> a: int = 1
>>> print(a)
1
```

**Comment by mattip** [comment]
This reminded me of unfinished #3984, where the conclusion was that overriding a node with another node should warn.

**Comment by da-woods** [comment]
> Fair enough but in this case the code should print `1` not `1.0`.

You're right - I missed that detail.

The general view we take on annotations is "they're not written specifically for Cython so Cython should use them if possible but not error if they're confusing". That's different from a "cdef" statement which is definitely written for Cython.

So my view is that `a` should ideally end up untyped in your example. But it looks like that isn't what actually happens.

**Comment by 0dminnimda** [comment]
> I think it is better if cython will be more strict than python -> not allowing redeclaring variables rather then generating code which behaves in a way that user does not expect.

I and probably others always thought of cython as some kind of superset of python, is it a good reson to break this?
It feels like cython always tries to follow the idea if it's like that in python, we do it the same way

**Comment by scoder** [comment]
> > I think it is better if cython will be more strict than python -> not allowing redeclaring variables rather then generating code which behaves in a way that user does not expect.
> 
> I and probably others always thought of cython as some kind of superset of python, is it a good reson to break this? It feels like cython always tries to follow the idea if it's like that in python, we do it the same way

[PEP 526](https://peps.python.org/pep-0526/#global-and-local-variable-annotations) has this to say on the matter:

```
Duplicate type annotations will be ignored. However, static type checkers may issue a warning for annotations of the same variable by a different type:

a: int
a: str  # Static type checker may or may not warn about this.
```

So, in Python, this is a warning at best.

However, when you run Cython on typed code, you are not just saying, "please compile this Python code for me", you are also explicitly enabling static typing. I think that explicit step of a user from "run this Python code with type annotations ignored at runtime" to "compile this Python code with static typing" is enough of a reason to let Cython reject contradicting type declarations with a compile error. This should not just be a warning, even if Cython does not completely understand the annotation yet.

If we encounter conflicting type annotations in user code, it seems more likely (to me) that it's a bug than intention. Intuitively, if you want different types for a variable, then why do you use the same variable name in the first place? The differently typed cases are probably really referring to different things that are worth having different names.

**Comment by da-woods** [comment]
This nice thing about @scoder's interpretation (vs mine) is that it's probably much easier to implement the error message for it than it'd be to diagnose multiple conflicting annotations and revert back to untyped.

**Comment by matusvalo** [comment]
I personally prefer @scoder solution.
